js/views/conversation_view.js

```diff
@@ -577,33 +580,77 @@
       //   events up to its parent elements in the DOM.
       this.closeMenu();
 
-      const media = await Signal.Backbone.Conversation.fetchVisualMediaAttachments({
-        conversationId: this.model.get('id'),
-        WhisperMessageCollection: Whisper.MessageCollection,
+      // We fetch more documents than media as they don’t require to be loaded
+      // into memory right away. Revisit this once we have infinite scrolling:
+      const DEFAULT_MEDIA_FETCH_COUNT = 50;
+      const DEFAULT_DOCUMENTS_FETCH_COUNT = 150;
+
+      const conversationId = this.model.get('id');
+      const WhisperMessageCollection = Whisper.MessageCollection;
+      const rawMedia = await Signal.Backbone.Conversation.fetchVisualMediaAttachments({
+        conversationId,
+        count: DEFAULT_MEDIA_FETCH_COUNT,
+        WhisperMessageCollection,
+      });
+      const documents = await Signal.Backbone.Conversation.fetchFileAttachments({
+        conversationId,
+        count: DEFAULT_DOCUMENTS_FETCH_COUNT,
+        WhisperMessageCollection,
       });
-      const loadMessages = Signal.Components.PropTypes.Message
+
+      // NOTE: Could we show grid previews from disk as well?
+      const loadMessages = Signal.Components.Types.Message
         .loadWithObjectURL(Signal.Migrations.loadMessage);
-      const mediaWithObjectURLs = await loadMessages(media);
-
-      const mediaGalleryProps = {
-        media: mediaWithObjectURLs,
-        documents: [],
-        onItemClick: ({ message }) => {
-          const lightboxProps = {
-            imageURL: message.objectURL,
-          };
-          this.lightboxView = new Whisper.ReactWrapperView({
-            Component: Signal.Components.Lightbox,
-            props: lightboxProps,
-            onClose: () => Signal.Backbone.Views.Lightbox.hide(),
-          });
-          Signal.Backbone.Views.Lightbox.show(this.lightboxView.el);
-        },
+      const media = await loadMessages(rawMedia);
+
+      const saveAttachment = async ({ message } = {}) => {
+        const attachment = message.attachments[0];
+        const timestamp = message.received_at;
+        Signal.Types.Attachment.save({
+          attachment,
+          document,
+          getAbsolutePath: Signal.Migrations.getAbsoluteAttachmentPath,
+          timestamp,
+        });
+      };
+
+      const onItemClick = async ({ message, type }) => {
+        switch (type) {
+          case 'documents': {
+            saveAttachment({ message });
+            break;
+          }
+
+          case 'media': {
+            const selectedIndex = media.findIndex(mediaMessage =>
+              mediaMessage.id === message.id);
+            const { getAbsoluteAttachmentPath } = Signal.Migrations;
+            this.lightboxGalleryView = new Whisper.ReactWrapperView({
+              Component: Signal.Components.LightboxGallery,
+              props: {
+                getAbsoluteAttachmentPath,
```

Comment:
Interesting. I think I'd like these kinds of things to be in selectors and not called directly in the react component. That would argue for some sort of pre-processing of these items before they go into the react component, of course. More work, but it would set a good precedent architecturally.